### DISCUSSION TITLE
How to batch multiple store updates?

### DISCUSSION THREAD

**Discussion by raviteja83** [pmndrs/zustand#2275] [Q&A]
I have a websocket where I receive messages and store them in a zustand store. Sometimes, I would receive more than 40msg/sec or so. For each message I am converting the object to a particular format and storing in zustand store. In this case, 40 setState have happened. Each updating the same fields(adding to the existing list).  Also, I am rendering the above messages in a virtualised list
How do I improve the handling of this and reduce the number of setState calls? After sometime, the UI is frozen and CPU usage goes to 100%. What can I do to prevent the UI freeze?

**Comment by dai-shi** [comment]
For frequent updates, we shouldn't trigger React for each update.

One technique is known as transient update.
https://github.com/pmndrs/zustand/blob/517524d2b05843c6a26d2713af08cc0ca1f62f3c/readme.md?plain=1#L247-L262

There might be another way, but we shouldn't use `useStore` directly in all such cases.

---
If anyone has similar experience, don't hesitate to share ideas.

**Reply by raviteja83** [reply]
I have used transient in other places, here I have to render the message as well. Need to figure out a way to do this with transient updates(I am using virtualised list here). I was trying to batch that event, for some interval, seems to be better performant but still need to more testing. Any other ideas are welcome.

**Reply by dai-shi** [reply]
another idea would be batching updates outside of Zustand.

**Comment by gabrr** [comment]
I would put those updates in a queue, save them in an object, within 60/120 seconds an agent would be responsible to check that object and set the data to the you’re using currently — that’s important so your whole UI don’t keep being updated.


Best regards,
Gabriel Oliveira
+55 48 99107 5872
